### PR TITLE
The `UHUGEINT` type

### PR DESCRIPTION
**Pull request by nickgerrets**
This PR adds the `UHUGEINT` type, an unsigned variant of the `HUGEINT` type. The type is also known as `UINT128`, an unsigned 128-bit integer.

`UHUGEINT` contains almost all features that `HUGEINT` also has, including:
- Casts from/to all integral types.
- Casts from/to string (`VARCHAR`).
- Casts from/to floating point numbers.
- Casts from/to `DECIMAL` type.
- Casts from/to `BIT` type.
- Arithmetic.
- Overflow checking (no wrapping).
- Bitwise operations.
- Support for in-tree extensions.

Considering there is no integral type larger than this type, arithmetic between `UHUGEINT` and `HUGEINT` will "promote" the type to double, potentially causing a loss in precision if the values are large.

The minimum value of `UHUGEINT` is: `0`.
The maximum value of `UHUGEINT` is: `340282366920938463463374607431768211455`.

EDIT: There is NO support for numpy (yet).

### CODE REVIEW

File: src/common/arrow/arrow_converter.cpp [hunk 1]
```diff
@@ -123,6 +123,9 @@ void SetArrowFormat(DuckDBArrowSchemaHolder &root_holder, ArrowSchema &child, co
 	case LogicalTypeId::HUGEINT:
 		child.format = "d:38,0";
 		break;
+	case LogicalTypeId::UHUGEINT:
+		child.format = "d:38,0";
```
Comment (Tishj) [inline src/common/arrow/arrow_converter.cpp:127]: I wonder if this might be a problem
treating hugeint and uhugeint the same in arrow

File: test/sql/storage/compression/bitpacking/bitpacking_types.test_slow [hunk 2]
```diff
@@ -13,7 +13,7 @@ foreach bitpacking_mode delta_for for constant_delta constant
 statement ok
 PRAGMA force_bitpacking_mode='${bitpacking_mode}'
 
-foreach type <numeric> hugeint decimal(4,1) decimal(8,1) decimal(12,1) decimal(18,1)
```
Comment (Tishj) [inline test/sql/storage/compression/bitpacking/bitpacking_types.test_slow:16]: What is the reason `hugeint` was removed here?

File: tools/jdbc/src/main/java/org/duckdb/DuckDBResultSet.java [hunk 3]
```diff
@@ -1242,6 +1249,8 @@ public <T> T getObject(int columnIndex, Class<T> type) throws SQLException {
             if (sqlType == DuckDBColumnType.HUGEINT) {
                 throw new SQLException("Can't convert value to BigInteger " + type.toString());
                 // return type.cast(getLocalDateTime(columnIndex));
+            } else if (sqlType == DuckDBColumnType.UHUGEINT) {
+                throw new SQLException("Can't convert valye to BigInteger " + type.toString());
```
Comment (Tishj) [inline tools/jdbc/src/main/java/org/duckdb/DuckDBResultSet.java:1253]: `valye` -> `value`

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java [hunk 4]
```diff
@@ -967,12 +967,13 @@ public static void test_duckdb_getObject_with_class() throws Exception {
...
                      + " dec32 DECIMAL(9,8), dec64 DECIMAL(16,1), dec128 DECIMAL(30,10), tint TINYINT, utint UTINYINT,"
-                     + " usint USMALLINT, uint UINTEGER, ubig UBIGINT, hin HUGEINT, blo BLOB)");
+                     + " usint USMALLINT, uint UINTEGER, ubig UBIGINT, hin HUGEINT, uhin UHUGEINT, blo BLOB)");
         stmt.execute(
             "INSERT INTO b VALUES ('varchary', true, 6, 42, 666, 42.666, 666.42,"
             +
             " '1970-01-02', '01:00:34', '1970-01-03 03:42:23', 42.2, 1.23456789, 987654321012345.6, 111112222233333.44444, "
-            + " -4, 200, 50001, 4000111222, 18446744073709551615, 18446744073709551616, 'yeah'::BLOB)");
+            +
+            " -4, 200, 50001, 4000111222, 18446744073709551615, 18446744073709551616, 18446744073709551616, 'yeah'::BLOB)");
```
Comment (Tishj) [inline tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:976]: I think `18446744073709551616` is used here for HUGEINT because it surpasses the max of UBIGINT, I believe `170141183460469231731687303715884105728` would surpass the max for HUGEINT so this would be a better test here for UHUGEINT

**Review by Tishj — COMMENTED** [review]
This is truly a mountain of work, thanks for the effort and the time you've spent to support this new addition!

I am slightly worried that because this is such a mountain of work, and touches all of our clients and connectors, from the C api, to Julia, Arrow, Parquet, Python, Swift, JDBC and possibly more.. that it is going to be quite hard to confidently say this is safe to merge.

You have added a ton of tests for the changes you've made though, so I might just be overly pessimistic

Maybe in the future it would be better to try to split this up and start off with throwing some NotImplementedExceptions

Never the less I think overall the PR looks great

File: test/sql/storage/compression/bitpacking/bitpacking_types.test_slow (hunk 2 above)
Comment (nickgerrets) [reply test/sql/storage/compression/bitpacking/bitpacking_types.test_slow:16]: `hugeint` is already part of `numeric`.

File: src/common/arrow/arrow_converter.cpp (hunk 1 above)
Comment (nickgerrets) [reply src/common/arrow/arrow_converter.cpp:129]: Looking into this a bit more, I've discovered that treating `UHUGEINT` as a `DECIMAL` type does indeed cause some issues as it treats the value as signed. For `HUGEINT` that's mostly okay, even if there is no conversion back to `HUGEINT` from arrow, because the values are pretty much the same. There doesn't appear to be a really good way to handle this. However, there are two more obvious solutions:
- Set the arrow format for `UHUGEINT` to be the same as `DOUBLE` ("g").
- Remove `UHUGEINT` from the supported Arrow types.

Either way there is no clear way of converting back from Arrow-schema to `UHUGEINT` or `HUGEINT`.

Comment (Tishj) [reply src/common/arrow/arrow_converter.cpp:129]: I think removing UHUGEINT from the supported Arrow types for now sounds like the best option, throwing a NotImplementedException

**Comment by l1t1** [comment]
clickhouse has int256 too.
https://clickhouse.com/docs/en/sql-reference/data-types/int-uint

**Comment by nickgerrets** [comment]
PR now up to date with the 0.9.0 release.
Due to being too similar to the `HUGEINT` implementation within Python (numpy), support for numpy has been dropped and usage will result in NotImplementedException.

**Comment by Mytherin** [comment]
Could you merge this with feature?

File: test/sql/function/list/list_concat.test [hunk 5]
```diff
@@ -61,7 +61,7 @@ foreach type_a <integral> varchar
 foreach type_b <integral> varchar
 
 query T
-SELECT list_concat([1, 2]::${type_a}[], [3, 4]::${type_b}[])
+SELECT list_concat([1, 2]::${type_a}[], [3, 4]::${type_b}[])::INTEGER[]
```
Comment (Mytherin) [inline test/sql/function/list/list_concat.test:64]: Why is this necessary?

File: test/sql/storage/types/test_uhugeint_storage.test [hunk 6]
```diff
@@ -0,0 +1,37 @@
+# name: test/sql/storage/types/test_uhugeint_storage.test
+# description: Test UHUGEINT with persistent storage
+# group: [types]
+
+# load the DB from disk
+load __TEST_DIR__/uhugeint_storage_test.db
+
+# create a table with uhugeints
+statement ok
+CREATE TABLE uhugeints (h HUGEINT);
```
Comment (Mytherin) [inline test/sql/storage/types/test_uhugeint_storage.test:10]: Can we turn this into `UHUGEINT` and also test the min/max values of the type?

File: test/sql/types/numeric/uhugeint_try_cast.test [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+statement ok
+PRAGMA enable_verification
+
+# integer casts
+statement ok
+CREATE TABLE uhugeints AS SELECT i::UHUGEINT i FROM (VALUES (0::UHUGEINT), (1::UHUGEINT), ('340282366920938463463374607431768211455'::UHUGEINT)) tbl(i)
+
+# -> signed
+statement error
+SELECT i::TINYINT FROM uhugeints
```
Comment (Mytherin) [inline test/sql/types/numeric/uhugeint_try_cast.test:14]: Can we check the error messages here?

File: test/sql/types/uhugeint/test_uhugeint_exponent.test [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+54321
+
+query I
+select (0.00000000000000000000002e+44)::UHUGEINT;
+----
+2000000000000000000000
+
+# overflow
+statement error
+select '340282366920938463463374607431768211456e0'::UHUGEINT
```
Comment (Mytherin) [inline test/sql/types/uhugeint/test_uhugeint_exponent.test:30]: Can we check the error messages here?

File: test/sql/types/uhugeint/test_uhugeint_exponent.test [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+statement error
+select '1.7e39'::UHUGEINT
```
Comment (Mytherin) [inline test/sql/types/uhugeint/test_uhugeint_exponent.test:33]: Should this be out-of-range?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great - some minor comments:

File: test/sql/function/list/list_concat.test (hunk 5 above)
Comment (nickgerrets) [reply test/sql/function/list/list_concat.test:64]: When integers `type_a` is signed and `type_b` is unsigned, the types in the list get promoted to a larger size signed integer. Considering `HUGEINT` and `UHUGEINT` have no larger type to promote to, they promote to `DOUBLE` which when printed will include the the decimal dot and 0. Casting to a list of integers is purely done to prevent that.

```
SELECT list_concat([1, 2]::hugeint[], [3, 4]::uhugeint[]);
================================================================================
Mismatch on row 1, column 1
[1.0, 2.0, 3.0, 4.0] <> [1, 2, 3, 4]
```

Comment (Mytherin) [reply test/sql/function/list/list_concat.test:64]: Aha, thanks, makes sense

File: test/sql/types/uhugeint/test_uhugeint_exponent.test (hunk 9 above)
Comment (nickgerrets) [reply test/sql/types/uhugeint/test_uhugeint_exponent.test:33]: Yes, `3.4e38` is approaching the maximum of `UHUGEINT`. `3.5e38` will overflow, therefore `1.7e39` will as well.

**Comment by Mytherin** [comment]
Thanks!
